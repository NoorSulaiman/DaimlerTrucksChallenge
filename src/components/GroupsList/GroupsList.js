import React, { useState, useEffect } from "react";
import { FormControl, Select, MenuItem } from "@mui/material";

const GroupsList = ({ components, selectGroup }) => {
  const [group, setGroup] = useState("");
  const [groupList, setGroupList] = useState([]);
  useEffect(() => {
    sortGroups();
  }, []);
  const sortGroups = () => {
    //this function will take componentes location prefixes and sort them in array without duplicates.
    //used object to store prefixes for duplicate check and filtering efficiency
    let groupStore = {};
    let sortedGroupArray = [];
    if (components.length) {
      components.map((component) => {
        let prefix = component.location.slice(0, 2);
        if (!groupStore[prefix]) {
          groupStore[prefix] = prefix;
          sortedGroupArray.push(prefix);
        }
      });
    }
    setGroupList(sortedGroupArray.sort());
  };
  const handleGroupChange = (event) => {
    setGroup(event.target.value);
    selectGroup(event.target.value);
  };
  const menuItems = groupList.map((prefix) => {
    return (
      <MenuItem key={prefix} value={prefix}>
        {prefix}
      </MenuItem>
    );
  });
  return (
    <div>
      <FormControl sx={{ m: 1, minWidth: 120 }}>
        <Select
          value={group}
          onChange={handleGroupChange}
          displayEmpty
          inputProps={{ "aria-label": "Without label" }}>
          <MenuItem value="">
            <em>Select Group Prefix...</em>
          </MenuItem>
          {menuItems}
        </Select>
      </FormControl>
    </div>
  );
};

export default GroupsList;
