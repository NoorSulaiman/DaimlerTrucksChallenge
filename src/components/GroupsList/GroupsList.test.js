import React from "react";
import { shallow } from "enzyme";
import GroupsList from "./GroupsList";

const clickFn = jest.fn();
const compArray = [
  {
    panelNumber: "22-58883-000",
    panelDescription: "PNL-INR TRIMPLTKAHLUA SPRKL",
    location: "AGAAA",
    parentPanelNumber: "734-C02202",
    modelCode: "H",
    whitelist: ["AGAAA"],
    blacklist: [],
    type: "hardwired-component",
    switchId: "",
    bom: "734-C02202",
    isAsBuilt: true,
  },
  {
    panelNumber: "22-69410-000",
    panelDescription: "PNL-SW,SMARTPLX,OVHD,SD,KAHLUA",
    location: "OVLHP",
    parentPanelNumber: "679-C01232",
    modelCode: "H",
    whitelist: ["OVLHP", "OVRHP"],
    blacklist: [],
    type: "hardwired-component",
    switchId: "",
    bom: "679-C01232",
    isAsBuilt: true,
  },
];
const sortGroups = (components) => {
  let groupStore = {};
  let sortedGroupArray = [];
  if (components.length) {
    components.map((component) => {
      let prefix = component.location.slice(0, 2);
      if (!groupStore[prefix]) {
        groupStore[prefix] = prefix;
        sortedGroupArray.push(prefix);
      }
    });
  }
  return sortedGroupArray.sort();
};
describe("GroupsList", () => {
  it("should render GroupsList correctly with given props", () => {
    const component = shallow(
      <GroupsList components={compArray} selectGroup={clickFn} />
    );
    expect(component).toMatchSnapshot();
  });

  it("should sort components list into group prefixes", () => {
    const component = shallow(
      <GroupsList components={compArray} selectGroup={clickFn} />
    );
    expect(component.find("Select").children.length).toBe(1);
  });

  it("should take prefixs and sort them", () => {
    expect(sortGroups(compArray)).toStrictEqual(["AG", "OV"]);
  });
});
