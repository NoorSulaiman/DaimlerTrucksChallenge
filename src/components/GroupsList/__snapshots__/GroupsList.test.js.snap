// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupsList should render GroupsList correctly with given props 1`] = `ShallowWrapper {}`;
